File: recipes/java-recipes/spring-boot-recipe/src/main/java/com/trendyol/stove/examples/java/spring/infra/boilerplate/kafka/KafkaDomainEventPublisher.java
package com.trendyol.stove.examples.java.spring.infra.boilerplate.kafka;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.trendyol.stove.examples.domain.ddd.AggregateRoot;
import com.trendyol.stove.examples.domain.ddd.EventPublisher;
import java.util.stream.Stream;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.slf4j.Logger;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Component;

@Component
public class KafkaDomainEventPublisher implements EventPublisher {
  private final KafkaTemplate<String, Object> template;
  private final TopicResolver topicResolver;
  private final ObjectMapper objectMapper;
  private final Logger logger = org.slf4j.LoggerFactory.getLogger(KafkaDomainEventPublisher.class);

  public KafkaDomainEventPublisher(
      KafkaTemplate<String, Object> template,
      TopicResolver topicResolver,
      @Qualifier("objectMapper") ObjectMapper objectMapper) {
    this.template = template;
    this.topicResolver = topicResolver;
    this.objectMapper = objectMapper;
  }

  @Override
  public <TId> void publishFor(AggregateRoot<TId> aggregateRoot) {
    mapEventsToProducerRecords(aggregateRoot).forEach(template::send);
  }

  private <TId> Stream<ProducerRecord<String, Object>> mapEventsToProducerRecords(
      AggregateRoot<TId> aggregateRoot) {
    return aggregateRoot.domainEvents().stream()
        .map(
            event -> {
              var topic = topicResolver.resolve(aggregateRoot.getAggregateName());
              try {
                logger.info("Publishing event {} to topic {}", event, topic.getName());
                return new ProducerRecord<>(
                    topic.getName(),
                    aggregateRoot.getIdAsString(),
                    objectMapper.writeValueAsString(event));
              } catch (JsonProcessingException e) {
                throw new RuntimeException(e);
              }
            });
  }
}

File: recipes/kotlin-recipes/ktor-recipe/src/main/kotlin/com/trendyol/stove/examples/kotlin/ktor/infra/boilerplate/kafka/KafkaDomainEventPublisher.kt
package com.trendyol.stove.examples.kotlin.ktor.infra.boilerplate.kafka

import com.fasterxml.jackson.databind.ObjectMapper
import com.trendyol.stove.examples.domain.ddd.*
import io.github.nomisRev.kafka.publisher.KafkaPublisher
import kotlinx.coroutines.runBlocking
import org.apache.kafka.clients.producer.ProducerRecord
import org.slf4j.*

class KafkaDomainEventPublisher(
  private val publisher: KafkaPublisher<String, Any>,
  private val topicResolver: TopicResolver,
  private val objectMapper: ObjectMapper
) : EventPublisher {
  private val logger: Logger = LoggerFactory.getLogger(KafkaDomainEventPublisher::class.java)

  override fun <TId> publishFor(aggregateRoot: AggregateRoot<TId>) = runBlocking {
    mapEventsToProducerRecords(aggregateRoot)
      .forEach { record -> publisher.publishScope { offer(record) } }
  }

  private fun <TId> mapEventsToProducerRecords(
    aggregateRoot: AggregateRoot<TId>
  ): List<ProducerRecord<String, Any>> = aggregateRoot.domainEvents()
    .map { event ->
      val topic: Topic = topicResolver(aggregateRoot.aggregateName)
      logger.info("Publishing event {} to topic {}", event, topic.name)
      ProducerRecord<String, Any>(
        topic.name,
        aggregateRoot.idAsString,
        objectMapper.writeValueAsString(event)
      )
    }
}
